src/easylink/pipeline_schema.py

```diff
@@ -1,45 +1,45 @@
 from pathlib import Path
-from typing import Dict, List, Optional
+from typing import Iterable
 
-import networkx as nx
 from layered_config_tree import LayeredConfigTree
 
+from easylink.graph_components import EdgeParams
 from easylink.pipeline_schema_constants import ALLOWED_SCHEMA_PARAMS
-from easylink.step import CompositeStep, Step
+from easylink.step import CompositeState, HierarchicalStep, Step
 
 
-class PipelineSchema(CompositeStep):
+class PipelineSchema(HierarchicalStep):
     """
-    A schema is a CompositeStep whose StephGraph determines all possible
+    A schema is a Step whose StephGraph determines all possible
     allowable pipelines.
     """
 
+    def __init__(self, name: str, nodes: Iterable[Step], edges: Iterable[EdgeParams]) -> None:
+        super().__init__(name, nodes=nodes, edges=edges)
+
     def __repr__(self) -> str:
         return f"PipelineSchema.{self.name}"
 
     def set_step_config(self, parent_config: LayeredConfigTree) -> None:
         self._config = parent_config
 
-    @property
-    def step_nodes(self) -> List[str]:
-        """Return list of nodes tied to specific steps."""
-        ordered_nodes = list(nx.topological_sort(self.step_graph))
-        return [node for node in ordered_nodes if node != "input_data" and node != "results"]
+    def validate_step(
+        self, pipeline_config: LayeredConfigTree, input_data_config: LayeredConfigTree
+    ) -> dict[str, list[str]]:
+        return super().validate_step({"substeps": pipeline_config}, input_data_config)
```

Comment:
I am pretty confident I understand this, but it could use a comment, e.g.

```python
# Treat the entire pipeline configuration as if it were nested under the "substeps" key of a root
# hierarchical step. This must be added here because this root step doesn't exist from the user's
# perspective -- it doesn't appear explicitly in the pipeline.yaml
```